libraries/freertos_plus/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,1829 @@
+{
+    size_t parsedBytes = http_parser_execute(pHttpParser, &_httpParserSettings, pBuf, len );
+    const char * pHttpParserErrorDescription = NULL;
+
+    /* If the parser fails with HPE_CLOSED_CONNECTION or HPE_INVALID_CONSTANT that simply means there
+       was data beyond the end of the message. We do not fail in this case because we often give the whole
+       header buffer or body buffer to the parser even if it is only partly filled with data. 
+       HPE_CB_message_begin <= error <= HPE_CB_chunk_complete covers all errors for returning a non-zer value 
+       from the parser callbacks. I return a nonzero from the callback when I want to stop the parser early like
+       for a HEAD request or when we find part of the response body in the header buffer. */
+    if(pHttpParser->http_errno != 0 &&
+        ( HTTP_PARSER_ERRNO( pHttpParser ) != HPE_CLOSED_CONNECTION ) &&
+        ( HTTP_PARSER_ERRNO( pHttpParser ) != HPE_INVALID_CONSTANT ) &&
+        ( ( HTTP_PARSER_ERRNO( pHttpParser ) >= HPE_CB_message_begin) &&
+          ( HTTP_PARSER_ERRNO( pHttpParser ) <= HPE_CB_chunk_complete) ))
+    {
+        pHttpParserErrorDescription = http_errno_description( HTTP_PARSER_ERRNO( pHttpParser ) );
+        IotLogError("http_parser failed on the http response with error: %s", pHttpParserErrorDescription);
+        return IOT_HTTPS_PARSING_ERROR;
+    }
+
+    return IOT_HTTPS_OK;
+}
+
+/*-----------------------------------------------------------*/
+
+static IotHttpsReturnCode_t _receiveHttpsMessage( _httpsConnection_t* _httpsConnection, 
+    http_parser* pParser,
+    IotHttpsResponseParserState_t *pCurrentParserState,
+    IotHttpsResponseParserState_t finalParserState, 
+    uint8_t** pBuf,
+    uint8_t** pBufCur,
+    uint8_t** pBufEnd,
+    IotHttpsReturnCode_t *pNetworkStatus)
+{    
+    IotHttpsReturnCode_t status = IOT_HTTPS_OK;
+
+    while( (*pCurrentParserState < finalParserState) && (*pBufEnd - *pBufCur > 0) ) 
+    {
+        *pNetworkStatus = _networkRecv( _httpsConnection,
+            *pBufCur, 
+            *pBufEnd - *pBufCur);
+
+        status = _parseHttpsMessage(pParser, *pBufCur, *pBufEnd - *pBufCur);
+        if(status != IOT_HTTPS_OK)
+        {
+            IotLogError("Failed to parse the message buffer with error: %d", pParser->http_errno);
+            return status;
+        }
+
+        /* The _httResponse->pHeadersCur pointer is updated in the http_parser callbacks. */
+        IotLogDebug( "There is %d of space left in the buffer.", *pBufEnd - *pBufCur );
+
+        /* We cannot anticipate the unique network error received when the server closes the connection. 
+           We simply exit the loop if there nothing else to receive. We do not return the network status because 
+           the error may just be a server closed the connection, but there may still have been data in the buffer
+           that we parsed. */
+        if( *pNetworkStatus != IOT_HTTPS_OK )
```

Comment:
No because I want to still parse the data received in the case of an error. If there is an error there may still be HTTP data.

Reply:
```
        /* We cannot anticipate the unique network error received when the server closes the connection. 
           We simply exit the loop if there nothing else to receive. We do not return the network status because 
           the error may just be a server closed the connection, but there may still have been HTTP data in the buffer
           that we parsed. */
```